### ISSUE TITLE
Handling users' blank lines

### ISSUE THREAD

**Issue by clement26695** [jhipster/prettier-java#164]
In some cases, blank line(s) are added by the user for clarity and readability to separate code into logical subsections.

I believe we should keep these blank lines, but I don't think we can handle that with the current implementation of the parser. WDYT ?

**Comment by bd82** [comment]
> but I don't think we can handle that with the current implementation of the parser

What data are we missing to handle this?

**Comment by clement26695** [comment]
For the moment, blank lines are not considered as a token/node, and so they does not appear in the tree. For instance, if we have this kind of code : 

```java
public class Foo {
    private Integer group1Value = 1;
    private List<Entity> group1;
    
    private String group2Name = "group2";
    private Integer group2Value = 2;
    private List<Entity> group2;
}
```
We will have a list of 5 "fieldDeclaration", without any information of a blank lines at line 4. In the printer, it will lead to 

```java
public class Foo {
    private Integer group1Value = 1;
    private List<Entity> group1;
    private String group2Name = "group2";
    private Integer group2Value = 2;
    private List<Entity> group2;
}
```
which is not suitable, IMO.

One way to detect this blank line migth be to compare the endLine/startLine of each "fieldDeclaration", if this information is available in the future

**Comment by bd82** [comment]
> One way to detect this blank line migth be to compare the endLine/startLine of each "fieldDeclaration", if this information is available in the future

This information is already available just not in an easy to consume manner...
This Chevrotain open issue could resolve the information accessibility 
- https://github.com/SAP/chevrotain/issues/932

**Comment by Shaolans** [comment]
I think we can do it the same way we did for the comments, in the java-parser we do a post computing where on each token we add a field "trailingLines" giving the number of blank lines below the tokens.
We can then add the line in the getImageWithComments function. The annoying case might be blank lines for comments but I don't think it is too complicated.
If this seems good to you, I will probably to it.

**Comment by bd82** [comment]
I think that this may be slightly different than comments.
Comments can appear anywhere, but while blank lines can also appear anywhere
we dont need to respect those anywhere.

e.g:
```javascript
var x = 1 +
   2 + 
   
   3 // the blank line above should not be respected
```

So maybe there is a limited number of scenarios where blank links should be respected.
Lets say between statements.

Also we may allow only a limited additional number of blank lines (e.g 2 lines between statements max, but never 20 blank lines), need to see how prettier handles this in EcmaScript

**Comment by Shaolans** [comment]
Oh you are totally right, I totally forgot about that kind of cases :smile:

**Comment by clement26695** [comment]
Closed with #202
